nixos/modules/services/security/kanidm.nix

```
@@ -0,0 +1,335 @@
+let
+  cfg = config.services.kanidm;
+  settingsFormat = pkgs.formats.toml { };
+  serverConfigFile = settingsFormat.generate "server.toml" cfg.serverSettings;
+  clientConfigFile = settingsFormat.generate "kanidm-config.toml" cfg.clientSettings;
+  unixConfigFile = settingsFormat.generate "kanidm-unixd.toml" cfg.unixSettings;
+
+  defaultServiceConfig = {
+    BindReadOnlyPaths = [
+      "/nix/store"
+      "-/etc/resolv.conf"
+      "-/etc/nsswitch.conf"
+      "-/etc/hosts"
+      "-/etc/localtime"
+    ];
+    CapabilityBoundingSet = "";
+    # ProtectClock= adds DeviceAllow=char-rtc r
+    DeviceAllow = "";
+    # Implies ProtectSystem=strict, which re-mounts all paths
+    #DynamicUser = true;
+    LockPersonality = true;
+    MemoryDenyWriteExecute = true;
+    NoNewPrivileges = true;
+    PrivateDevices = true;
+    PrivateMounts = true;
+    PrivateNetwork = true;
+    PrivateTmp = true;
+    PrivateUsers = true;
+    ProcSubset = "pid";
+    ProtectClock = true;
+    ProtectHome = true;
+    ProtectHostname = true;
+    # Would re-mount paths ignored by temporary root
+    #ProtectSystem = "strict";
+    ProtectControlGroups = true;
+    ProtectKernelLogs = true;
+    ProtectKernelModules = true;
+    ProtectKernelTunables = true;
+    ProtectProc = "invisible";
+    RestrictAddressFamilies = [ ];
+    RestrictNamespaces = true;
+    RestrictRealtime = true;
+    RestrictSUIDSGID = true;
+    SystemCallArchitectures = "native";
+    SystemCallFilter = [ "@system-service" "~@privileged @resources @setuid @keyring" ];
+    # Does not work well with the temporary root
+    #UMask = "0066";
+  };
+
+in
+{
+  options.services.kanidm = {
+    enable = lib.mkEnableOption "the kanidm server";
+    enablePam = lib.mkEnableOption "the pam and nsswitch integration of kanidm";
+
+    package = lib.mkOption {
+      type = lib.types.package;
+      default = pkgs.kanidm;
+      description = "Which kanidm package to use.";
+    };
+
+    ensureDomainName = lib.mkOption {
+      description = ''
+        The domain_name that Kanidm manages. It must be a descendant or equal to the domain
+        specified in settings.origin. It is always set when the server starts up.
+        While in theory it could be changed later, you should carefully consider these
+        warnings:
+        <link xlink:href="https://kanidm.github.io/kanidm/administrivia.html#rename-the-domain"/>
+      '';
+      example = "example.org";
+      type = lib.types.str;
+    };
+
+    serverSettings = lib.mkOption {
+      type = lib.types.submodule {
+        freeformType = settingsFormat.type;
+
+        options = {
+          bindaddress = lib.mkOption {
+            description = "Address/port combination the webserver binds to.";
+            example = "[::1]:8443";
+            type = lib.types.str;
+          };
+          ldapbindaddress = lib.mkOption {
+            description = "Address/port combination the emulated ldap server binds to.";
```

```suggestion
            description = ''
              Address and port the LDAP server is bound to. Setting this to an empty string disables the LDAP interface.
            '';
```

Better explains the behaviour of the default value.